File: Cavern/Filters/Utilities/FilterGraphNodeUtils.cs
using System.Collections.Generic;
using System.Linq;

namespace Cavern.Filters.Utilities {
    /// <summary>
    /// Special functions for handling <see cref="FilterGraphNode"/>s.
    /// </summary>
    public static partial class FilterGraphNodeUtils {
        /// <summary>
        /// Convert the filter graph's filters to convolutions, and merge chains together to a single filter.
        /// </summary>
        public static void ConvertToConvolution(this IEnumerable<IFilterGraphNode> rootNodes, int sampleRate, int filterLength) {
            HashSet<IFilterGraphNode> visited = new HashSet<IFilterGraphNode>();
            Queue<IFilterGraphNode> queue = new Queue<IFilterGraphNode>(rootNodes);
            while (queue.Count > 0) {
                IFilterGraphNode currentNode = queue.Dequeue();
                if (visited.Contains(currentNode)) {
                    continue;
                }

                if (!(currentNode.Filter is BypassFilter || currentNode.Filter.GetType() == typeof(FastConvolver))) {
                    ConvertToConvolution(currentNode, sampleRate, filterLength);
                }
                visited.Add(currentNode);
                foreach (IFilterGraphNode child in currentNode.Children) {
                    queue.Enqueue(child);
                }
            }
        }

        /// <summary>
        /// Creates a copy of the complete graph with no overlapping memory with the old <paramref name="rootNodes"/>.
        /// </summary>
        public static IEnumerable<IFilterGraphNode> DeepCopy(this IEnumerable<IFilterGraphNode> rootNodes) =>
            DeepCopyWithMapping(rootNodes).rootNodes;

        /// <summary>
        /// Creates a copy of the complete graph with no overlapping memory with the old <paramref name="rootNodes"/>,
        /// and also results which old root <see cref="IFilterGraphNode"/> maps to which new one.
        /// </summary>
        public static (IEnumerable<IFilterGraphNode> rootNodes, Dictionary<IFilterGraphNode, IFilterGraphNode> mapping)
            DeepCopyWithMapping(this IEnumerable<IFilterGraphNode> rootNodes) {
            Dictionary<IFilterGraphNode, IFilterGraphNode> mapping = new Dictionary<IFilterGraphNode, IFilterGraphNode>();

            IFilterGraphNode CopyNode(IFilterGraphNode source) {
                if (mapping.ContainsKey(source)) {
                    return mapping[source];
                }

                IFilterGraphNode copy = (IFilterGraphNode)source.Clone();
                mapping[source] = copy;
                foreach (IFilterGraphNode child in source.Children) {
                    copy.AddChild(CopyNode(child));
                }
                return copy;
            }

            List<IFilterGraphNode> result = new List<IFilterGraphNode>();
            foreach (IFilterGraphNode rootNode in rootNodes) {
                result.Add(CopyNode(rootNode));
            }
            return (result, mapping);
        }

        /// <summary>
        /// Check if the graph has cycles.
        /// </summary>
        /// <param name="rootNodes">All nodes which have no parents</param>
        public static bool HasCycles(this IEnumerable<IFilterGraphNode> rootNodes) {
            HashSet<IFilterGraphNode> visited = new HashSet<IFilterGraphNode>(),
                inProgress = new HashSet<IFilterGraphNode>();
            foreach (IFilterGraphNode node in rootNodes) {
                if (!visited.Contains(node)) {
                    if (HasCycles(node, visited, inProgress)) {
                        return true;
                    }
                }
            }
            return false;
        }

        /// <summary>
        /// Converts this filter to a convolution and upmerges all children until possible.
        /// </summary>
        static void ConvertToConvolution(IFilterGraphNode node, int sampleRate, int filterLength) {
            float[] impulse = new float[filterLength];
            impulse[0] = 1;

            IFilterGraphNode downmergeUntil = node;
            while (true) {
                downmergeUntil.Filter.Process(impulse);
                if (downmergeUntil.Children.Count != 1 || downmergeUntil.Children[0].Parents.Count != 1 ||
                    downmergeUntil.Children[0].Filter is BypassFilter) {
                    break;
                }
                downmergeUntil = downmergeUntil.Children[0];
            }

            IFilterGraphNode[] newChildren = downmergeUntil.Children.ToArray();
            downmergeUntil.DetachChildren();
            node.Filter = new FastConvolver(impulse, sampleRate, 0);
            node.DetachChildren();
            for (int i = 0; i < newChildren.Length; i++) {
                node.AddChild(newChildren[i]);
            }
        }

        /// <summary>
        /// Starting from a single node, checks if the graph has cycles.
        /// </summary>
        static bool HasCycles(IFilterGraphNode currentNode, HashSet<IFilterGraphNode> visited, HashSet<IFilterGraphNode> inProgress) {
            if (inProgress.Contains(currentNode)) {
                return true;
            }
            if (visited.Contains(currentNode)) {
                return false;
            }

            inProgress.Add(currentNode);
            foreach (IFilterGraphNode child in currentNode.Children) {
                if (HasCycles(child, visited, inProgress)) {
                    return true;
                }
            }
            inProgress.Remove(currentNode);
            visited.Add(currentNode);
            return false;
        }
    }
}

File: CavernAmp/Cavern/Filters/Utilities/filterGraphNodeUtils.cpp
#include <algorithm>
#include <cstring>

#include "filterGraphNodeUtils.h"

std::vector<FilterGraphNode*> FilterGraphNodeUtils::DeepCopy(
    const std::vector<FilterGraphNode*>& rootNodes) {
    auto result = DeepCopyWithMapping(rootNodes);
    return result.first;
}

std::pair<std::vector<FilterGraphNode*>, std::unordered_map<FilterGraphNode*, FilterGraphNode*>>
FilterGraphNodeUtils::DeepCopyWithMapping(
    const std::vector<FilterGraphNode*>& rootNodes) {
    std::unordered_map<FilterGraphNode*, FilterGraphNode*> mapping;

    std::function<FilterGraphNode*(FilterGraphNode*)> copyNode;
    copyNode = [&](FilterGraphNode* source) -> FilterGraphNode* {
        auto it = mapping.find(source);
        if (it != mapping.end()) {
            return it->second;
        }

        FilterGraphNode* copy = new FilterGraphNode(*source);
        mapping[source] = copy;
        for (size_t i = 0; i < source->GetChildren().size(); i++) {
            copy->AddChild(copyNode(source->GetChildren()[i]));
        }
        return copy;
    };

    std::vector<FilterGraphNode*> result;
    for (size_t i = 0; i < rootNodes.size(); i++) {
        result.push_back(copyNode(rootNodes[i]));
    }
    return std::make_pair(result, mapping);
}

bool FilterGraphNodeUtils::HasCycles(const std::vector<FilterGraphNode*>& rootNodes) {
    std::unordered_set<FilterGraphNode*> visited;
    std::unordered_set<FilterGraphNode*> inProgress;
    for (size_t i = 0; i < rootNodes.size(); i++) {
        if (visited.find(rootNodes[i]) == visited.end()) {
            if (HasCycles(rootNodes[i], visited, inProgress)) {
                return true;
            }
        }
    }
    return false;
}

bool FilterGraphNodeUtils::HasCycles(
    FilterGraphNode* currentNode,
    std::unordered_set<FilterGraphNode*>& visited,
    std::unordered_set<FilterGraphNode*>& inProgress) {
    if (inProgress.find(currentNode) != inProgress.end()) {
        return true;
    }
    if (visited.find(currentNode) != visited.end()) {
        return false;
    }

    inProgress.insert(currentNode);
    const auto& children = currentNode->GetChildren();
    for (size_t i = 0; i < children.size(); i++) {
        if (HasCycles(children[i], visited, inProgress)) {
            return true;
        }
    }
    inProgress.erase(currentNode);
    visited.insert(currentNode);
    return false;
}

std::unordered_set<FilterGraphNode*> FilterGraphNodeUtils::MapGraph(
    const std::vector<FilterGraphNode*>& rootNodes,
    std::function<const std::vector<FilterGraphNode*>& (FilterGraphNode*)> direction) {
    std::unordered_set<FilterGraphNode*> visited;
    std::queue<FilterGraphNode*> queue;
    for (size_t i = 0; i < rootNodes.size(); i++) {
        queue.push(rootNodes[i]);
    }

    while (!queue.empty()) {
        FilterGraphNode* currentNode = queue.front();
        queue.pop();

        if (visited.find(currentNode) != visited.end()) {
            continue;
        }

        visited.insert(currentNode);
        const auto& nextSteps = direction(currentNode);
        for (size_t i = 0; i < nextSteps.size(); i++) {
            queue.push(nextSteps[i]);
        }
    }

    return visited;
}

std::unordered_set<FilterGraphNode*> FilterGraphNodeUtils::MapGraph(
    const std::vector<FilterGraphNode*>& rootNodes) {
    return MapGraph(rootNodes, [](FilterGraphNode* node) -> const std::vector<FilterGraphNode*>& {
        return node->GetChildren();
    });
}

std::unordered_set<FilterGraphNode*> FilterGraphNodeUtils::MapGraphBack(
    const std::vector<FilterGraphNode*>& rootNodes) {
    return MapGraph(rootNodes, [](FilterGraphNode* node) -> const std::vector<FilterGraphNode*>& {
        return node->GetParents();
    });
}

bool FilterGraphNodeUtils::IsTopologicalSort(const std::vector<FilterGraphNode*>& orderedNodes) {
    std::unordered_set<FilterGraphNode*> visited;
    for (size_t i = 0; i < orderedNodes.size(); i++) {
        FilterGraphNode* node = orderedNodes[i];
        const auto& children = node->GetChildren();
        for (size_t j = 0; j < children.size(); j++) {
            if (visited.find(children[j]) != visited.end()) {
                return false;
            }
        }
        visited.insert(node);
    }
    return true;
}

std::vector<FilterGraphNode*> FilterGraphNodeUtils::TopologicalSort(
    const std::vector<FilterGraphNode*>& rootNodes) {
    std::vector<FilterGraphNode*> result;
    std::unordered_set<FilterGraphNode*> visited;

    std::function<void(FilterGraphNode*)> visitNode;
    visitNode = [&](FilterGraphNode* node) {
        if (visited.find(node) != visited.end()) {
            return;
        }
        visited.insert(node);
        const auto& children = node->GetChildren();
        for (size_t i = 0; i < children.size(); i++) {
            visitNode(children[i]);
        }
        result.insert(result.begin(), node);
    };

    for (size_t i = 0; i < rootNodes.size(); i++) {
        visitNode(rootNodes[i]);
    }
    return result;
}

void FilterGraphNodeUtils::ConvertToConvolution(
    FilterGraphNode* node, int sampleRate, int filterLength) {
    float* impulse = new float[filterLength]();
    impulse[0] = 1;

    FilterGraphNode* downmergeUntil = node;
    while (true) {
        downmergeUntil->pFilter->Process(impulse, filterLength);
        const auto& children = downmergeUntil->GetChildren();
        if (children.size() != 1 || downmergeUntil->GetParents().size() != 1) {
            break;
        }
        downmergeUntil = children[0];
    }

    const auto& newChildrenList = downmergeUntil->GetChildren();
    std::vector<FilterGraphNode*> newChildren(newChildrenList.begin(), newChildrenList.end());
    downmergeUntil->DetachChildren();

    node->pFilter = new FastConvolver(impulse, filterLength, 0);
    delete[] impulse;

    node->DetachChildren();
    for (size_t i = 0; i < newChildren.size(); i++) {
        node->AddChild(newChildren[i]);
    }
}

void FilterGraphNodeUtils::ConvertToConvolution(
    const std::vector<FilterGraphNode*>& rootNodes, int sampleRate, int filterLength) {
    std::unordered_set<FilterGraphNode*> visited;
    std::queue<FilterGraphNode*> queue;
    for (size_t i = 0; i < rootNodes.size(); i++) {
        queue.push(rootNodes[i]);
    }

    while (!queue.empty()) {
        FilterGraphNode* currentNode = queue.front();
        queue.pop();

        if (visited.find(currentNode) != visited.end()) {
            continue;
        }

        ConvertToConvolution(currentNode, sampleRate, filterLength);
        visited.insert(currentNode);

        const auto& children = currentNode->GetChildren();
        for (size_t i = 0; i < children.size(); i++) {
            queue.push(children[i]);
        }
    }
}
